sotodlib/hwp/hwp.py

```diff
@@ -6,429 +6,524 @@
 
 
-def func_hwpss_2f4f(x, A2, A4, phi2, phi4):
-    return A2 * np.cos(2 * x + phi2) + A4 * np.cos(4 * x + phi4)
+def extract_hwpss(aman, signal=None, hwp_angle=None,
+                  bin_signal=True, bins=3600,
+                  lin_reg=True, modes=[1, 2, 3, 4, 6, 8],
+                  apply_prefilt=True, prefilt_cutoff=1.0, prefilt_detrend='linear',
+                  flags=None,
+                  merge_stats=True, hwpss_stats_name='hwpss_stats',
+                  merge_extract=True, hwpss_extract_name='hwpss_extract'):
+    """
+    Extracts HWP synchronous signal (HWPSS) from a time-ordered data (TOD) using linear regression or curve-fitting.
+
+    Parameters
+    ----------
+    aman : AxisManager object
+        The TOD to extract HWPSS from.
+    signal : array-like, optional
+        The TOD signal to use. If not provided, `aman.signal` will be used.
+    hwp_angle : array-like, optional
+        The HWP angle for each sample in `aman`. If not provided, `aman.hwp_angle` will be used.
+    bin_signal : bool, optional
+        Whether to bin the TOD signal into HWP angle bins before extracting HWPSS. Default is `True`.
+    bins : int, optional
+        The number of HWP angle bins to use if `bin_signal` is `True`. Default is 3600.
+    lin_reg : bool, optional
+        Whether to use linear regression to extract HWPSS from the binned signal. If `False`, curve-fitting will be used instead.
+        Default is `True`.
+    modes : list of int, optional
+        The HWPSS harmonic modes to extract. Default is [1, 2, 3, 4, 6, 8].
+    apply_prefilt : bool, optional
+        Whether to apply a high-pass filter to signal before extracting HWPSS. Default is `True`.
+    prefilt_cutoff : float, optional
+        The cutoff frequency of the high-pass filter, in Hz. Only used if `apply_prefilt` is `True`. Default is 1.0.
+    prefilt_detrend: str or None
+        Method of detrending when you apply prefilter. Default is `linear`. If data is already detrended or you do not want to detrend,
+        set it to `None`.
+    flags : RangesMatrix, optional
+        Flags to be masked out before extracting HWPSS. If Default is None, and no mask will be applied.
+    merge_stats : bool, optional
+        Whether to add the extracted HWPSS statistics to `aman` as new axes. Default is `True`.
+    hwpss_stats_name : str, optional
+        The name to use for the new field containing the HWPSS statistics if `merge_stats` is `True`. Default is 'hwpss_stats'.
+    merge_extract : bool, optional
+        Whether to add the extracted HWPSS to `aman` as a new signal field. Default is `True`.
+    hwpss_extract_name : str, optional
+        The name to use for the new signal field containing the extracted HWPSS if `merge_extract` is `True`. Default is 'hwpss_extract'.
+
+    Returns
+    -------
+    hwpss_stats : AxisManager object
+        The extracted HWPSS and its statistics.
+    """
 
+    if signal is None:
+        if apply_prefilt:
```

Comment:
Ya this is largely because the `fourier_filter`, `fft_ops`, and `detrend` modules break from the convention I understand we're supposed to follow. The convention we're using here is that the function should have a signal argument that defaults to None and will use `signal = tod.signal` in that case otherwise it expects a ndarray to be passed to signal. In the other modules they take a `signal_name` which is one of the field names in the axis manager. I went through and fixed this in a copy of each of those three modules on my machine but it touches a lot code so I was wondering if I should open that as a different PR?

Reply:
Yeah, that has confused me. I vote to the different PR plan.

Reply:
For now, then, apply_prefilt=True and signal!=None should raise an error.

I think we might need to initiate a transitional form for fourier_filter and friends, where they accept either signal_name or signal (but not both!).  Then gradually change all users over.

Reply:
yeah, I'm coming to the same conclusion (but we have a previous discussion https://github.com/simonsobs/sotodlib/issues/79 where you mentioned specifically seeing it was useful and I'm trying to remember _why_)

Reply:
I added a ValueError in this [commit ](https://github.com/simonsobs/sotodlib/pull/437/commits/e1bdf52c4e4f39d7d41a3ccd8684e3435ed1ad6a) so I'll resolve this comment with that